File: sdk/c/src/providers/fakemail.c
/**
 * FakeMail -- https://www.fakemail.net
 */

#include "tempmail_internal.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define FAKEMAIL_BASE "https://www.fakemail.net"

static const char *fakemail_home_headers[] = {
    "Accept: text/html,application/xhtml+xml",
    "User-Agent: Mozilla/5.0",
    NULL
};

static char *fakemail_merge_cookie(const char *prev, const char *set_cookie) {
    if (!set_cookie || !set_cookie[0]) return tm_strdup(prev ? prev : "");
    const char *end = strchr(set_cookie, ';');
    size_t n = end ? (size_t)(end - set_cookie) : strlen(set_cookie);
    if (!prev || !prev[0]) {
        char *out = (char*)malloc(n + 1);
        if (!out) return NULL;
        memcpy(out, set_cookie, n);
        out[n] = '\0';
        return out;
    }
    size_t cap = strlen(prev) + n + 3;
    char *out = (char*)malloc(cap);
    if (!out) return NULL;
    snprintf(out, cap, "%s; %.*s", prev, (int)n, set_cookie);
    return out;
}

static char *fakemail_csrf(const char *html) {
    const char *needle = "const CSRF=\"";
    const char *p = html ? strstr(html, needle) : NULL;
    if (!p) return tm_strdup("");
    p += strlen(needle);
    const char *end = strchr(p, '"');
    if (!end) return tm_strdup("");
    size_t n = (size_t)(end - p);
    char *out = (char*)malloc(n + 1);
    if (!out) return NULL;
    memcpy(out, p, n);
    out[n] = '\0';
    return out;
}

static const char *fakemail_json_start(const char *body) {
    if (!body) return "";
    const unsigned char *p = (const unsigned char*)body;
    if (p[0] == 0xef && p[1] == 0xbb && p[2] == 0xbf) return body + 3;
    return body;
}

static char *fakemail_urlencode(const char *s) {
    if (!s) return tm_strdup("");
    size_t len = strlen(s);
    char *out = (char*)malloc(len * 3 + 1);
    if (!out) return NULL;
    char *p = out;
    static const char hex[] = "0123456789ABCDEF";
    for (size_t i = 0; i < len; i++) {
        unsigned char c = (unsigned char)s[i];
        if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') ||
            (c >= '0' && c <= '9') || c == '-' || c == '_' || c == '.' || c == '~') {
            *p++ = (char)c;
        } else {
            *p++ = '%';
            *p++ = hex[c >> 4];
            *p++ = hex[c & 15];
        }
    }
    *p = '\0';
    return out;
}

static cJSON *fakemail_ajax_json(const char *path, const char *cookie, const char *post_body) {
    size_t cookie_cap = strlen(cookie) + 9;
    char *cookie_header = (char*)malloc(cookie_cap);
    if (!cookie_header) return NULL;
    snprintf(cookie_header, cookie_cap, "Cookie: %s", cookie);
    const char *headers[] = {
        "Accept: application/json, text/javascript, */*; q=0.01",
        "X-Requested-With: XMLHttpRequest",
        "Referer: https://www.fakemail.net/",
        "User-Agent: Mozilla/5.0",
        cookie_header,
        post_body ? "Content-Type: application/x-www-form-urlencoded" : NULL,
        NULL
    };
    char url[512];
    snprintf(url, sizeof(url), "%s%s", FAKEMAIL_BASE, path);
    tm_http_response_t *resp = tm_http_request_ipv4(post_body ? TM_HTTP_POST : TM_HTTP_GET, url, headers, post_body, 15);
    free(cookie_header);
    if (!resp || resp->status < 200 || resp->status >= 300) {
        tm_http_response_free(resp);
        return NULL;
    }
    cJSON *root = cJSON_Parse(fakemail_json_start(resp->body));
    tm_http_response_free(resp);
    return root;
}

tm_email_info_t* tm_provider_fakemail_generate(void) {
    tm_http_response_t *home = tm_http_request_ipv4(TM_HTTP_GET, FAKEMAIL_BASE "/", fakemail_home_headers, NULL, 15);
    if (!home || home->status < 200 || home->status >= 300) {
        tm_http_response_free(home);
        return NULL;
    }
    char *csrf = fakemail_csrf(home->body);
    char *cookie = fakemail_merge_cookie("", home->cookies);
    tm_http_response_free(home);
    if (!csrf || !csrf[0] || !cookie || !cookie[0]) {
        free(csrf);
        free(cookie);
        return NULL;
    }
    char *csrf_enc = fakemail_urlencode(csrf);
    free(csrf);
    if (!csrf_enc) {
        free(cookie);
        return NULL;
    }
    size_t path_cap = strlen(csrf_enc) + 32;
    char *path = (char*)malloc(path_cap);
    if (!path) {
        free(csrf_enc);
        free(cookie);
        return NULL;
    }
    snprintf(path, path_cap, "/index/index?csrf_token=%s", csrf_enc);
    free(csrf_enc);
    cJSON *root = fakemail_ajax_json(path, cookie, NULL);
    free(path);
    if (!root) {
        free(cookie);
        return NULL;
    }
    const char *email = TM_JSON_STR(cJSON_GetObjectItemCaseSensitive(root, "email"), "");
    if (!email[0] || !strchr(email, '@')) {
        cJSON_Delete(root);
        free(cookie);
        return NULL;
    }
    tm_email_info_t *info = tm_email_info_new();
    if (!info) {
        cJSON_Delete(root);
        free(cookie);
        return NULL;
    }
    info->channel = CHANNEL_FAKEMAIL;
    info->email = tm_strdup(email);
    info->token = cookie;
    cJSON_Delete(root);
    return info;
}

static cJSON *fakemail_fetch_detail(const char *cookie, const char *id) {
    char *id_enc = fakemail_urlencode(id);
    if (!id_enc) return NULL;
    size_t body_cap = strlen(id_enc) + 4;
    char *body = (char*)malloc(body_cap);
    if (!body) {
        free(id_enc);
        return NULL;
    }
    snprintf(body, body_cap, "id=%s", id_enc);
    free(id_enc);
    cJSON *root = fakemail_ajax_json("/index/email", cookie, body);
    free(body);
    return root;
}

static cJSON *fakemail_message_raw(const cJSON *row, const cJSON *detail, const char *recipient) {
    cJSON *raw = cJSON_CreateObject();
    if (!raw) return NULL;
    const cJSON *id = detail ? cJSON_GetObjectItemCaseSensitive((cJSON*)detail, "id") : NULL;
    if (!id) id = cJSON_GetObjectItemCaseSensitive((cJSON*)row, "id");
    const cJSON *from = detail ? cJSON_GetObjectItemCaseSensitive((cJSON*)detail, "od") : NULL;
    if (!from) from = cJSON_GetObjectItemCaseSensitive((cJSON*)row, "od");
    const cJSON *subject = detail ? cJSON_GetObjectItemCaseSensitive((cJSON*)detail, "predmet") : NULL;
    if (!subject) subject = cJSON_GetObjectItemCaseSensitive((cJSON*)row, "predmet");
    const cJSON *html = detail ? cJSON_GetObjectItemCaseSensitive((cJSON*)detail, "telo") : NULL;
    cJSON_AddStringToObject(raw, "id", TM_JSON_STR(id, ""));
    cJSON_AddStringToObject(raw, "from", TM_JSON_STR(from, ""));
    cJSON_AddStringToObject(raw, "to", recipient);
    cJSON_AddStringToObject(raw, "subject", TM_JSON_STR(subject, ""));
    cJSON_AddStringToObject(raw, "text", "");
    cJSON_AddStringToObject(raw, "html", TM_JSON_STR(html, ""));
    cJSON_AddStringToObject(raw, "date", TM_JSON_STR(cJSON_GetObjectItemCaseSensitive((cJSON*)row, "kdy"), ""));
    const char *read_flag = TM_JSON_STR(cJSON_GetObjectItemCaseSensitive((cJSON*)row, "precteno"), "");
    cJSON_AddBoolToObject(raw, "isRead", strcmp(read_flag, "precteno") == 0);
    return raw;
}

tm_email_t* tm_provider_fakemail_get_emails(const char *token, const char *email, int *count) {
    *count = -1;
    if (!token || !token[0] || !email || !email[0]) return NULL;
    cJSON *rows = fakemail_ajax_json("/index/refresh", token, NULL);
    if (!rows || !cJSON_IsArray(rows)) {
        cJSON_Delete(rows);
        return NULL;
    }
    int n = cJSON_GetArraySize(rows);
    *count = n;
    if (n == 0) {
        cJSON_Delete(rows);
        return NULL;
    }
    tm_email_t *emails = tm_emails_new(n);
    if (!emails) {
        *count = -1;
        cJSON_Delete(rows);
        return NULL;
    }
    for (int i = 0; i < n; i++) {
        cJSON *row = cJSON_GetArrayItem(rows, i);
        const char *id = TM_JSON_STR(cJSON_GetObjectItemCaseSensitive(row, "id"), "");
        cJSON *detail = id[0] ? fakemail_fetch_detail(token, id) : NULL;
        cJSON *raw = fakemail_message_raw(row, detail, email);
        emails[i] = tm_normalize_email(raw ? raw : row, email);
        if (raw) cJSON_Delete(raw);
        if (detail) cJSON_Delete(detail);
    }
    cJSON_Delete(rows);
    return emails;
}

File: sdk/rust/src/providers/fakemail.rs
/*!
 * FakeMail -- https://www.fakemail.net
 */

use crate::config::{block_on, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};
use regex::Regex;
use serde::Deserialize;
use serde_json::{Map, Value};
use std::sync::OnceLock;
use wreq::header::HeaderMap;

const BASE_URL: &str = "https://www.fakemail.net";

#[derive(Debug, Deserialize)]
struct GenerateResponse {
    email: Option<String>,
}

fn csrf_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r#"const\s+CSRF\s*=\s*"([^"]+)""#).unwrap())
}

fn merge_cookie(prev: &str, headers: &HeaderMap) -> String {
    let mut jar: Vec<(String, String)> = Vec::new();
    for part in prev.split(';') {
        let p = part.trim();
        if let Some((k, v)) = p.split_once('=') {
            if !k.is_empty() {
                jar.push((k.to_string(), v.to_string()));
            }
        }
    }
    for value in headers.get_all("set-cookie").iter() {
        let Ok(line) = value.to_str() else { continue };
        let first = line.split(';').next().unwrap_or("").trim();
        let Some((k, v)) = first.split_once('=') else {
            continue;
        };
        if let Some(pos) = jar.iter().position(|(name, _)| name == k) {
            jar[pos].1 = v.to_string();
        } else {
            jar.push((k.to_string(), v.to_string()));
        }
    }
    jar.into_iter()
        .map(|(k, v)| format!("{}={}", k, v))
        .collect::<Vec<_>>()
        .join("; ")
}

fn clean_json(text: &str) -> &str {
    text.strip_prefix('\u{feff}').unwrap_or(text)
}

fn open_session() -> Result<(String, String), String> {
    block_on(async {
        let resp = http_client()
            .get(format!("{}/", BASE_URL))
            .header("Accept", "text/html,application/xhtml+xml")
            .header("User-Agent", "Mozilla/5.0")
            .send()
            .await
            .map_err(|e| e.to_string())?;
        if !resp.status().is_success() {
            return Err(format!("fakemail home: {}", resp.status()));
        }
        let cookie = merge_cookie("", resp.headers());
        let html = resp.text().await.map_err(|e| e.to_string())?;
        let csrf = csrf_re()
            .captures(&html)
            .and_then(|caps| caps.get(1).map(|m| m.as_str().to_string()))
            .unwrap_or_default();
        if csrf.is_empty() {
            return Err("fakemail: csrf token not found".into());
        }
        Ok((csrf, cookie))
    })
}

fn create_address(csrf: &str, cookie: &str) -> Result<(GenerateResponse, String), String> {
    block_on(async {
        let resp = http_client()
            .get(format!(
                "{}/index/index?csrf_token={}",
                BASE_URL,
                urlencoding::encode(csrf)
            ))
            .header("Accept", "application/json, text/javascript, */*; q=0.01")
            .header("X-Requested-With", "XMLHttpRequest")
            .header("Referer", format!("{}/", BASE_URL))
            .header("Cookie", cookie)
            .header("User-Agent", "Mozilla/5.0")
            .send()
            .await
            .map_err(|e| e.to_string())?;
        if !resp.status().is_success() {
            return Err(format!("fakemail generate: {}", resp.status()));
        }
        let next_cookie = merge_cookie(cookie, resp.headers());
        let text = resp.text().await.map_err(|e| e.to_string())?;
        let data = serde_json::from_str::<GenerateResponse>(clean_json(&text))
            .map_err(|e| e.to_string())?;
        Ok((data, next_cookie))
    })
}

fn fetch_rows(cookie: &str) -> Result<Vec<Value>, String> {
    block_on(async {
        let resp = http_client()
            .get(format!("{}/index/refresh", BASE_URL))
            .header("Accept", "application/json, text/javascript, */*; q=0.01")
            .header("X-Requested-With", "XMLHttpRequest")
            .header("Referer", format!("{}/", BASE_URL))
            .header("Cookie", cookie)
            .header("User-Agent", "Mozilla/5.0")
            .send()
            .await
            .map_err(|e| e.to_string())?;
        if !resp.status().is_success() {
            return Err(format!("fakemail refresh: {}", resp.status()));
        }
        let text = resp.text().await.map_err(|e| e.to_string())?;
        serde_json::from_str::<Vec<Value>>(clean_json(&text)).map_err(|e| e.to_string())
    })
}

fn fetch_detail(cookie: &str, id: &str) -> Result<Value, String> {
    block_on(async {
        let resp = http_client()
            .post(format!("{}/index/email", BASE_URL))
            .header("Accept", "application/json, text/javascript, */*; q=0.01")
            .header("X-Requested-With", "XMLHttpRequest")
            .header("Referer", format!("{}/", BASE_URL))
            .header("Cookie", cookie)
            .header("Content-Type", "application/x-www-form-urlencoded")
            .header("User-Agent", "Mozilla/5.0")
            .form(&[("id", id)])
            .send()
            .await
            .map_err(|e| e.to_string())?;
        if !resp.status().is_success() {
            return Err(format!("fakemail detail: {}", resp.status()));
        }
        let text = resp.text().await.map_err(|e| e.to_string())?;
        serde_json::from_str::<Value>(clean_json(&text)).map_err(|e| e.to_string())
    })
}

fn flatten(row: &Value, detail: Option<&Value>, recipient: &str) -> Value {
    let detail_obj = detail.and_then(|x| x.as_object());
    let mut out = Map::new();
    out.insert(
        "id".into(),
        detail_obj
            .and_then(|m| m.get("id").cloned())
            .or_else(|| row.get("id").cloned())
            .unwrap_or(Value::String(String::new())),
    );
    out.insert(
        "from".into(),
        detail_obj
            .and_then(|m| m.get("od").cloned())
            .or_else(|| row.get("od").cloned())
            .unwrap_or(Value::String(String::new())),
    );
    out.insert("to".into(), Value::String(recipient.to_string()));
    out.insert(
        "subject".into(),
        detail_obj
            .and_then(|m| m.get("predmet").cloned())
            .or_else(|| row.get("predmet").cloned())
            .unwrap_or(Value::String(String::new())),
    );
    out.insert("text".into(), Value::String(String::new()));
    out.insert(
        "html".into(),
        detail_obj
            .and_then(|m| m.get("telo").cloned())
            .unwrap_or(Value::String(String::new())),
    );
    out.insert(
        "date".into(),
        row.get("kdy")
            .cloned()
            .unwrap_or(Value::String(String::new())),
    );
    out.insert(
        "isRead".into(),
        Value::Bool(row.get("precteno").and_then(|x| x.as_str()) == Some("precteno")),
    );
    Value::Object(out)
}

pub fn generate_email() -> Result<EmailInfo, String> {
    let (csrf, cookie) = open_session()?;
    let (data, cookie) = create_address(&csrf, &cookie)?;
    let email = data.email.unwrap_or_default().trim().to_string();
    if email.is_empty() || !email.contains('@') {
        return Err("fakemail: invalid mailbox response".into());
    }
    Ok(EmailInfo {
        channel: Channel::Fakemail,
        email,
        token: Some(cookie),
        expires_at: None,
        created_at: None,
    })
}

pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let cookie = token.trim();
    let address = email.trim();
    if cookie.is_empty() {
        return Err("fakemail: empty session token".into());
    }
    if address.is_empty() {
        return Err("fakemail: empty email".into());
    }
    let rows = fetch_rows(cookie)?;
    let mut out = Vec::new();
    for row in rows {
        let id = row.get("id").map(|x| x.to_string()).unwrap_or_default();
        let detail = if id.trim_matches('"').is_empty() {
            None
        } else {
            fetch_detail(cookie, id.trim_matches('"')).ok()
        };
        out.push(normalize_email(
            &flatten(&row, detail.as_ref(), address),
            address,
        ));
    }
    Ok(out)
}
